### PR TITLE
multi-component image slice rendering

### PR DESCRIPTION
**Pull request by thewtex**
@martinken not quite ready, but comments welcome.

### CODE REVIEW

**Comment by martinken** [comment]
Looks reasonable so far

**Comment by scottwittenburg** [comment]
@thewtex I ran the edited example (`TestVolumeTypes`) and it seems to kind of work, except at some point the slice view just becomes solid grey in conjunction with this error message:

```
WebGL warning: texImage2D: Desired upload requires more data than is available ...
```

Is that something you're aware of?  Additionally, can you briefly summarize what (else) you thought was left to finish on this PR?

**Comment by thewtex** [comment]
> Is that something you're aware of?

Yes, addressing this is the next step.

> Additionally, can you briefly summarize what (else) you thought was left to finish on this PR?

Better support for additional components, types, and the correct buffer / texture size is likely required somewhere here :-):

  https://github.com/Kitware/vtk-js/tree/master/Sources/Rendering/OpenGL

**Comment by thewtex** [comment]
Awesome @scottwittenburg !

I restarted the macOS CI build, and it passed :green_heart:

**Comment by scottwittenburg** [comment]
@thewtex This still needs a little history cleanup (I'll do it in the morning), but otherwise it's working as I demonstrated earlier today, and I've done the renaming/refactoring.  You may notice I changed my mind in that department again.  Let me know what you think.

**Comment by scottwittenburg** [comment]
Was it `npm run validate` that failed on travis?  That passes on my machine, and I can't find the actual cause of the error in the CI output.

**Comment by jourdain** [comment]
That's because I broke master. Someone needs to do the reformat.

You can fix it by rebasing and fixing it locally.

**Comment by scottwittenburg** [comment]
> That's because I broke master. Someone needs to do the reformat.

Ah, that explains why the push CI worked, and only the pr CI failed.  Thanks for the explanation.

> You can fix it by rebasing and fixing it locally.

Roger that, will do.  Thanks again.

**Comment by scottwittenburg** [comment]
@thewtex Ready for review

File: Examples/Volume/TestVolumeTypes/index.js [hunk 1]
```diff
@@ -2,25 +2,49 @@ import 'vtk.js/Sources/favicon';
...
-const renderWindow = vtkRenderWindow.newInstance();
-const renderer = vtkRenderer.newInstance();
-renderWindow.addRenderer(renderer);
+const fullScreenRenderWindow = vtkFullScreenRenderWindow.newInstance({
+  background: [0, 0, 0],
+  containerStyle: style,
+});
+const renderWindow = fullScreenRenderWindow.getRenderWindow();
+const renderer = fullScreenRenderWindow.getRenderer();
+// const interactor = fullScreenRenderWindow.getInteractor();
```
Comment (thewtex) [inline Examples/Volume/TestVolumeTypes/index.js]: Commented code

**Comment by scottwittenburg** [comment]
There's currently some kind of outage on Travis CI, which may explain why those checks disappeared.  Once it's back up, I can force push to trigger them to run.

File: Sources/Rendering/OpenGL/ImageMapper/index.js [hunk 2]
```diff
@@ -465,36 +598,64 @@ function vtkOpenGLImageMapper(publicAPI, model) {
...
+          } else {
+            for (let i = 0; i < cWidth * 3; i++) {
+              cTable[c * cWidth * 6 + i] = 255.0 * tmpTable[i];
+            }
+          }
         }
         model.colorTextureString = cfunToString;
+        if (!iComps && numComp === 1) {
+          console.log('Issues?');
+        }
```
Comment (thewtex) [inline Sources/Rendering/OpenGL/ImageMapper/index.js]: debug code

File: Sources/Rendering/OpenGL/ImageMapper/index.js [hunk 3]
```diff
@@ -465,36 +598,64 @@ function vtkOpenGLImageMapper(publicAPI, model) {
...
+        console.log(cWidth * 3);
+        console.log(textureHeight);
+        console.log(cTable.length);
```
Comment (thewtex) [inline Sources/Rendering/OpenGL/ImageMapper/index.js]: debug code

File: Examples/Volume/TestVolumeTypes/index.js [hunk 4]
```diff
@@ -2,25 +2,48 @@ import 'vtk.js/Sources/favicon';
...
 container.appendChild(renderWindowContainer);
 
+const style = {
+  margin: '0',
+  padding: '0',
+  position: 'absolute',
+  top: '0',
+  left: '0',
+  // width: '450px',
+  // height: '450px',
```
Comment (thewtex) [inline Examples/Volume/TestVolumeTypes/index.js]: debug code

File: Sources/Rendering/OpenGL/ImageMapper/index.js [hunk 5]
```diff
@@ -465,36 +598,64 @@ function vtkOpenGLImageMapper(publicAPI, model) {
...
         model.colorTexture.create2DFromRaw(
           cWidth,
-          1,
+          textureHeight,
```
Comment (thewtex) [inline Sources/Rendering/OpenGL/ImageMapper/index.js]: :+1:

File: Sources/Rendering/OpenGL/ImageMapper/index.js [hunk 6]
```diff
@@ -699,6 +876,10 @@ function vtkOpenGLImageMapper(publicAPI, model) {
         vtkErrorMacro('Reformat slicing not yet supported.');
       }
 
+      console.log(dims);
+      console.log(numComp);
+      console.log(scalars.length);
+      console.log(dims[0] * dims[1] * numComp);
```
Comment (thewtex) [inline Sources/Rendering/OpenGL/ImageMapper/index.js]: debug code

File: Sources/Rendering/OpenGL/ImageMapper/index.js [hunk 7]
```diff
@@ -519,39 +680,56 @@ function vtkOpenGLImageMapper(publicAPI, model) {
...
-    const ofun = actor.getProperty().getScalarOpacity();
-    if (ofun) {
-      const ofunToString = `${ofun.getMTime()}`;
-      if (model.opacityTextureString !== ofunToString) {
-        const oRange = ofun.getRange();
-        const ofTable = new Float32Array(oWidth);
-        ofun.getTable(oRange[0], oRange[1], oWidth, ofTable, 1);
-        for (let i = 0; i < oWidth; ++i) {
-          oTable[i] = 255.0 * ofTable[i];
+    // Build pwf buffer, it's either for comp weighting or opacity
```
Comment (thewtex) [inline Sources/Rendering/OpenGL/ImageMapper/index.js]: pwf -> piecewise weighting function
comp -> component

**Comment by scottwittenburg** [comment]
Thanks for the review @thewtex, by the way, it seems Travis CI is still experience some kind of outage:

https://www.traviscistatus.com/

So I guess we may need to wait to merge this until it's back up and we can see those checks pass.

**Comment by scottwittenburg** [comment]
So the `travis-ci/pr` check passed, and that's the key one I think since it tests the branch merged with master.  Maybe we're all set?

**Comment by thewtex** [comment]
Yes! :100:

**Comment by agirault** [comment]
This PR introduced a lot of `console.log` everywhere, is that on purpose or should they be removed?

**Comment by scottwittenburg** [comment]
Yes, I guess some made it through accidentally.  I'll make a clean up pass at it eventually, unless you get to it first.

**Comment by agirault** [comment]
No problem. I didn't know if somehow babel/webpack or something would automatically get rid of them or make them no-op when not in debug, which is why I asked. I'll probably take a stab at it at some point. Thanks.

**Pull request merged**